Add suspend and resume functionality

mdedetrich opened the pull request:
# About this change - What it does

This PR is currently an in progress PR on the suspend/resume functionality. Its currently blocked because of the following issues 

- https://discuss.lightbend.com/t/committing-a-kafka-cursor-after-chunk-upload-on-multipart-s3-upload/8941
- https://github.com/akka/alpakka/issues/2760
- https://github.com/akka/alpakka/pull/2770
- https://github.com/aiven/guardian-for-apache-kafka/issues/27 (we ended up having to use a killswitch to properly test this functionality however in production you should be using the controlled shutdown)

mdedetrich commented on core/src/main/scala/io/aiven/guardian/kafka/KafkaClient.scala:
```diff
@@ -43,7 +44,9 @@ class KafkaClient(
     val base = ConsumerSettings(system, new ByteArrayDeserializer, new ByteArrayDeserializer)
-    configure.fold(base)(block => block(base))
+    configure.fold(base)(block => block(base)).withProperties(
+      ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG -> false.toString
```
We want to have full control of committing the Kafka cursors, if this isn't set to false than Kafka will periodically do commits in backround (outside of our control) in addition to our manual commits after an S3 chunk is uploaded

mdedetrich commented on core/src/main/scala/io/aiven/guardian/kafka/KafkaClient.scala:
```diff
@@ -43,7 +44,11 @@ class KafkaClient(
+    configure
+      .fold(base)(block => block(base))
+      .withProperties(
+        ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG -> false.toString
```
We want to have full control of committing the Kafka cursors, if this isn't set to false than Kafka will periodically do commits in backround (outside of our control) in addition to our manual commits after an S3 chunk is uploaded

mdedetrich commented on core-backup/src/main/scala/io/aiven/guardian/kafka/backup/BackupClientInterface.scala:
```diff
@@ -247,18 +271,37 @@ trait BackupClientInterface[T <: KafkaClientInterface] {
 
+  /** Prepares the sink before it gets handed to `backupToStorageSink`
+    */
+  private[backup] def prepareStartOfStream(state: Option[CurrentState],
```
This function handles the case where if there is current state (i.e. you are resuming from a previously upload) then we need to drop the first character `[` from the stream since there is an already existing half complete upload which is the middle of an array

mdedetrich commented on core/src/main/scala/io/aiven/guardian/kafka/KafkaClient.scala:
```diff
@@ -14,44 +15,58 @@ import akka.stream.scaladsl.SourceWithContext
+    configureConsumer
+      .fold(base)(block => block(base))
+      .withProperties(
+        ConsumerConfig.AUTO_OFFSET_RESET_CONFIG -> "earliest"
```
This is the magic configuration without which the suspend/resume case does not work.

mdedetrich commented:
The PR is ready to review, the main notable change since the last comment is that the configuration for the timeslice has been changed (see `TimeConfiguration`)

mdedetrich commented on core-backup/src/main/scala/io/aiven/guardian/kafka/backup/BackupClientInterface.scala:
```diff
@@ -369,9 +412,16 @@ object BackupClientInterface {
+  ): Long = {
+    val (period, finalInitialTime) = timeConfiguration match {
+      case PeriodFromFirst(duration) => (duration, initialTime)
+      case ChronoUnitSlice(chronoUnit) =>
```
If its configured for the suspend/resume case then we need to set the initial time to also be the start of the time unit

mdedetrich commented on core/src/main/resources/reference.conf:
```diff
@@ -28,11 +28,12 @@ akka.kafka.consumer = {
+    max-interval = 1 hour
     max-interval = ${?AKKA_KAFKA_COMMITTER_MAX_INTERVAL}
     parallelism = ${?AKKA_KAFKA_COMMITTER_PARALLELISM}
-    delivery = ${?AKKA_KAFKA_COMMITTER_DELIVERY}
```
These configurations are being removed because they are hardcoded by the Kafka client itself so its impossible to override.

mdedetrich commented on core/src/test/scala/io/aiven/guardian/kafka/MockedKafkaClientInterface.scala:
```diff
@@ -47,6 +51,14 @@ class MockedKafkaClientInterface(kafkaData: Source[ReducedConsumerRecord, NotUse
   /** @return
     *   A `Sink` that allows you to commit a `CursorContext` to Kafka to signify you have processed a message
     */
-  override def commitCursor: Sink[Long, Future[Done]] = Sink.foreach(cursor => committedOffsets ++ Iterable(cursor))
```
This change is the same as what was done in https://github.com/aiven/guardian-for-apache-kafka/pull/76, i.e. `committedOffsets ++ Iterable(cursor)` doesn't actually end up doing anything because `committedOffsets` is immutable so we are just throwing away the reference.

mdedetrich commented on backup-s3/src/main/scala/io/aiven/guardian/kafka/backup/s3/BackupClient.scala:
```diff
@@ -1,40 +1,168 @@
+                      )
+                      single
+                    case rest =>
+                      val last = rest.minBy(_.initiated)(Ordering[Instant].reverse)
```
Yup, nice pickup

reta commented on backup-gcs/src/main/scala/io/aiven/guardian/kafka/backup/gcs/BackupClient.scala:
```diff
@@ -15,22 +16,35 @@ import io.aiven.guardian.kafka.gcs.configs.{GCS => GCSConfig}
 import scala.concurrent.Future
 
+// TODO: GCS implementation currently does not work correctly because of inability of current GCS implementation in
+// Alpakka to allows us to commit Kafka cursor whenever chunks are uploaded
```
Super minor: `to allows` -> `to allow`

reta commented:
LGTM but certainly @jlprat is needed here :-)

mdedetrich commented:
I just rebased the original commit, I forgot the change the `reference.conf` for the new `Backup` configuration. Ontop of this I also added tests (`ConfigSpec`) for this new configuration to make sure that the `reference.conf` is working.

jlprat commented on backup-s3/src/test/scala/io/aiven/guardian/kafka/backup/s3/RealS3BackupClientSpec.scala:
```diff
@@ -16,4 +56,289 @@ class RealS3BackupClientSpec extends AnyPropTestKit(ActorSystem("RealS3BackupCli
+    }
+  }
+
+  def waitUntilBackupClientHasCommitted(backupClient: BackupClientChunkState[_],
+                                        step: FiniteDuration = 100 millis,
+                                        delay: FiniteDuration = 5 seconds
+  ): Future[Unit] =
+    if (backupClient.processedChunks.size() > 0)
+      akka.pattern.after(delay)(Future.successful(()))
+    else
+      akka.pattern.after(step)(waitUntilBackupClientHasCommitted(backupClient, step, delay))
```
Nit pick, maybe move this auxiliary function on top, so all properties are together one after the other

mdedetrich replied:
Done

jlprat reviewed (approved):
I have a couple of minor comments, but LGTM

mdedetrich commented on backup-gcs/src/main/scala/io/aiven/guardian/kafka/backup/gcs/BackupClient.scala:
```diff
@@ -15,22 +16,35 @@ import io.aiven.guardian.kafka.gcs.configs.{GCS => GCSConfig}
   override type BackupResult = Option[StorageObject]
 
-  override def backupToStorageSink(key: String): Sink[ByteString, Future[BackupResult]] = {
+  override type CurrentState = Nothing
```
Correct, GCS is not implemented fully

mdedetrich commented on backup-s3/src/main/scala/io/aiven/guardian/kafka/backup/s3/BackupClient.scala:
```diff
@@ -1,40 +1,168 @@
 
+import java.time.Instant
+
+final case class CurrentS3State(uploadId: String, parts: Seq[Part])
```
Done

mdedetrich commented on backup-s3/src/main/scala/io/aiven/guardian/kafka/backup/s3/BackupClient.scala:
```diff
@@ -1,40 +1,168 @@
+  override type CurrentState = CurrentS3State
+
+  override def getCurrentUploadState(key: String): Future[Option[CurrentS3State]] = {
+    implicit val ec: ExecutionContext = system.classicSystem.getDispatcher
```
I also thought the same however for this case I believe its overcomplicated due to the fact that this ec will only get executed per time slice which in the grand scheme of things is nothing.

mdedetrich commented on build.sbt:
```diff
@@ -23,6 +23,7 @@ val akkaStreamsJson            = "0.8.0"
 val diffxVersion               = "0.5.6"
 val testContainersVersion      = "0.39.8"
 val testContainersJavaVersion  = "1.16.2"
+val scalaCheckVersion          = "1.15.5-1-SNAPSHOT"
```
It contains https://github.com/typelevel/scalacheck/pull/849 which is required to generated data that is approximately the size of S3 chunks

jlprat reviewed (approved):
LGTM
